src/combinatorics/hindman.lean

```diff
@@ -0,0 +1,143 @@
+          refine ⟨N_add _ _ mem'' mem', _⟩,
+          rw [set.mem_set_of_eq, add_assoc, eq', eq''], }, },
+      apply set.inter_subset_left, },
+    rw ←absorbing_eq at hm,
+    exact hm.2, },
+  apply zorn.zorn_superset,
+  intros c hcs hc,
+  refine ⟨⋂₀ c, ⟨is_closed_sInter $ λ t ht, (hcs ht).1, _, _⟩, _⟩,
+  { obtain rfl | hcnemp := c.eq_empty_or_nonempty,
+    { simp only [set.univ_nonempty, set.sInter_empty] },
+    convert @is_compact.nonempty_Inter_of_directed_nonempty_compact_closed _ _ _
+      (c.nonempty_coe_sort.mpr hcnemp) (coe : c → set M) _ _ _ _,
+    { simp only [subtype.range_coe_subtype, set.set_of_mem_eq] } ,
+    { intros a b,
+      cases hc.total_of_refl a.2 b.2,
+      { exact ⟨a, set.subset.rfl, h⟩, },
+      { exact ⟨b, h, set.subset.rfl⟩, }, },
+    { intro i, exact (hcs i.property).2.1, },
+    { intro i, exact is_closed.is_compact (hcs i.property).1, },
+    { intro i, exact (hcs i.property).1, }, },
+  { intros m m' hm hm',
+    rw set.mem_sInter,
+    intros t ht,
+    exact (hcs ht).2.2 m m' (set.mem_sInter.mp hm t ht) (set.mem_sInter.mp hm' t ht), },
+  { intros s hs, exact set.sInter_subset_of_mem hs, },
+end
+
+/-- `FS a m` says that `m : M` is the sum of a finite nonempty subsequence of `a : stream M`. We
+define it directly as an inductive prop instead of talking about finite subsequences. -/
+inductive FS {M} [add_semigroup M] : stream M → M → Prop
+| singleton (a : stream M) : FS a a.head
+| tail (a : stream M) (m : M) (h : FS a.tail m) : FS a m
+| add (a : stream M) (m : M) (h : FS a.tail m) : FS a (a.head + m)
+
+lemma exists_FS_of_large {M} [add_semigroup M] (U : ultrafilter M) (U_idem : U + U = U)
+  (s₀ : set M) (sU : s₀ ∈ U) : ∃ a, ∀ m, FS a m → m ∈ s₀ :=
+begin
+  have exists_elem : ∀ (s : set M) (hs : s ∈ U), (s ∩ { m | ∀ᶠ m' in U, m + m' ∈ s }).nonempty :=
+    λ s hs, ultrafilter.nonempty_of_mem (inter_mem hs $ by { rw ←U_idem at hs, exact hs }),
+  let elem : { s // s ∈ U } → M := λ p, set.nonempty.some (exists_elem p.val p.property),
+  let succ : { s // s ∈ U } → { s // s ∈ U } := λ p, ⟨p.val ∩ { m | elem p + m ∈ p.val }, _⟩,
+  use stream.corec elem succ (subtype.mk s₀ sU),
+  { suffices : ∀ (a : stream M) (m : M), FS a m → ∀ p, a = stream.corec elem succ p → m ∈ p.val,
+    { intros m hm, exact this _ m hm ⟨s₀, sU⟩ rfl, },
+    clear sU s₀,
+    intros a m h,
+    induction h with b b n h ih b n h ih,
+    { rintros p rfl,
+      rw [stream.corec_eq, stream.head_cons],
+      exact set.inter_subset_left _ _ (set.nonempty.some_mem _), },
+    { rintros p rfl,
+      refine set.inter_subset_left _ _ (ih (succ p) _),
+      rw [stream.corec_eq, stream.tail_cons], },
+    { rintros p rfl,
+      have := set.inter_subset_right _ _ (ih (succ p) _),
+      { simpa only using this },
+      rw [stream.corec_eq, stream.tail_cons], }, },
+  apply inter_mem p.property,
+  convert set.inter_subset_right _ _ (set.nonempty.some_mem $ exists_elem p.val p.property),
+end
+
+/-- **Hindman's theorem** -/
```

Comment:
Please add a bit more detail to the docstring. I really liked the module docstring at the top of this file.